### PR TITLE
(fix)O3-3046: Allow sorting vitals, biometrics and conditions rows based on different sort functions

### PR DESCRIPTION
**Pull request by vasharma05**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Summary
This PR adds custom sorting functions for every column in the vitals, biometrics and conditions table.

## Screenshots

https://github.com/openmrs/openmrs-esm-patient-chart/assets/51502471/e0bfb07c-73f8-4a9a-a6c3-53482268a854

## Related Issue
https://issues.openmrs.org/browse/O3-3046

## Other
None

### CODE REVIEW

**Review by ibacher — APPROVED** [review]
LGTM, but I'm unclear about why the conditions stuff is part of this PR?

File: packages/esm-patient-conditions-app/src/conditions/conditions.resource.ts [hunk 1]
```diff
@@ -233,3 +234,45 @@ export async function deleteCondition(conditionId: string) {
...
+}
+
+export interface ConditionTableHeader {
+  key: 'display' | 'onsetDateTimeRender' | 'status';
+  header: string;
+  isSortable: true;
+  sortFunc: (valueA: ConditionTableRow, valueB: ConditionTableRow) => number;
+}
+
+export function useConditionsSorting(tableHeaders: Array<ConditionTableHeader>, tableRows: Array<ConditionTableRow>) {
```
Comment (denniskigen) [inline packages/esm-patient-conditions-app/src/conditions/conditions.resource.ts:252]: Would something like `useSortedConditionTableRows` describe this hook better?

File: packages/esm-patient-vitals-app/src/biometrics/biometrics-base.component.tsx [hunk 2]
```diff
@@ -36,28 +37,50 @@ const BiometricsBase: React.FC<BiometricsBaseProps> = ({ patientUuid, pageSize,
...
 
-  const tableHeaders = [
-    { key: 'date', header: t('dateAndTime', 'Date and time'), isSortable: true },
-    { key: 'weight', header: withUnit(t('weight', 'Weight'), conceptUnits.get(config.concepts.weightUuid) ?? '') },
-    { key: 'height', header: withUnit(t('height', 'Height'), conceptUnits.get(config.concepts.heightUuid) ?? '') },
-    { key: 'bmi', header: `${t('bmi', 'BMI')} (${bmiUnit})` },
+  const tableHeaders: Array<BiometricsTableHeader> = [
     {
-      key: 'muac',
+      key: 'dateRender',
```
Comment (denniskigen) [inline packages/esm-patient-vitals-app/src/biometrics/biometrics-base.component.tsx:42]: What's the rationale behind the `Render` suffix?

File: packages/esm-patient-vitals-app/src/vitals/vitals-overview.component.tsx [hunk 3]
```diff
@@ -70,39 +71,69 @@ const VitalsOverview: React.FC<VitalsOverviewProps> = ({ patientUuid, pageSize,
     };
   }, [patient, t, excludePatientIdentifierCodeTypes?.uuids]);
 
-  const tableHeaders = [
-    { key: 'date', header: t('dateAndTime', 'Date and time'), isSortable: true },
+  const tableHeaders: Array<VitalsTableHeader> = [
     {
-      key: 'temperature',
+      key: 'dateRender',
```
Comment (denniskigen) [inline packages/esm-patient-vitals-app/src/vitals/vitals-overview.component.tsx:76]: Is the idea to repackage this logic as a custom hook, similar to what you did with conditions above?

**Review by denniskigen — APPROVED** [review]
A few concerns though it mostly looks good to me.

**Comment by denniskigen** [comment]
The failing vitals and biometrics tests are likely due to a change in the default sort order behaviour.

**Comment by vasharma05** [comment]
Fixed the failing tests @denniskigen

File: packages/esm-patient-vitals-app/src/biometrics/biometrics-base.component.tsx (hunk 2 above)
Comment (vasharma05) [reply packages/esm-patient-vitals-app/src/biometrics/biometrics-base.component.tsx:42]: There are 2 kinds of data for a column (say `date` column): 
1. `date`, which is the ISO string
2. `dateRender`, which is the formatted version of the pt.1 date.

Now, if we see, the user/ datatable sees the `dateRender` on the screen, but we want to sort the column based on the `date` value, hence I added the "Render" suffix.

File: packages/esm-patient-vitals-app/src/vitals/vitals-overview.component.tsx (hunk 3 above)
Comment (vasharma05) [reply packages/esm-patient-vitals-app/src/vitals/vitals-overview.component.tsx:76]: Yes @denniskigen, I'll move this hook inside the framework such that all the data tables can use this similar implementation.
The new need from the developer will be to add the `sortFunc` for every header in the tableHeaders.

If you have a better idea than this, I'm all open to hear it.
Thansk!

**Pull request merged**